tests/unit/test_correlation.py

```diff
@@ -0,0 +1,141 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Test to verify correlation ID functionality."""
+
+import asyncio
+import random
+from contextlib import nullcontext
+from contextvars import ContextVar
+
+import pytest
+
+from hexkit.correlation import (
+    CorrelationIdContextError,
+    InvalidCorrelationIdError,
+    correlation_id_var,
+    get_correlation_id,
+    new_correlation_id,
+    set_correlation_id,
+    validate_correlation_id,
+)
+from hexkit.utils import set_context_var
+
+
+async def set_id_sleep_resume(correlation_id: str, use_context_manager: bool):
+    """An async task to set the correlation ID ContextVar and yield control temporarily
+    back to the event loop before resuming.
+    """
+    if use_context_manager:
+        async with set_context_var(correlation_id_var, correlation_id):
+            await asyncio.sleep(random.random() * 2)  # Yield control to the event loop
+            # Check if the correlation ID is still the same
+            assert correlation_id_var.get() == correlation_id, "Correlation ID changed"
+    else:
+        token = correlation_id_var.set(correlation_id)  # Set correlation ID for task
+        await asyncio.sleep(random.random() * 2)  # Yield control to the event loop
+        # Check if the correlation ID is still the same
+        assert correlation_id_var.get() == correlation_id, "Correlation ID changed"
+        correlation_id_var.reset(token)
+
+
+@pytest.mark.asyncio
+async def test_correlation_id_isolation():
+    """Make sure correlation IDs are isolated to the respective async task and that
+    there's no interference from task switching.
+
+    Test with a sleep time of 0-2s and a random combination of context
+    manager/directly setting ContextVar.
+    """
+    tasks = [
+        set_id_sleep_resume(f"test_{n}", random.choice((True, False)))
+        for n in range(100)
+    ]
+    await asyncio.gather(*tasks)
+
+
+@pytest.mark.parametrize(
+    "correlation_id,exception",
+    [
+        ("BAD_ID", InvalidCorrelationIdError),
+        ("", InvalidCorrelationIdError),
+        (new_correlation_id(), None),
+    ],
+)
+@pytest.mark.asyncio
+async def test_correlation_id_validation(correlation_id: str, exception):
+    """Ensure an error is raised when correlation ID validation fails."""
+    with pytest.raises(exception) if exception else nullcontext():
+        validate_correlation_id(correlation_id)
+
+
+@pytest.mark.parametrize(
+    "correlation_id,exception",
+    [
+        ("12345", InvalidCorrelationIdError),
+        ("", InvalidCorrelationIdError),
+        (new_correlation_id(), None),
```

Comment:
Creating the new id is actually also non-deterministic, but I think we can live with it here.

Reply:
But maybe also add a fixed valid value before that line.